lib.rs

```diff
@@ -35,6 +35,104 @@ thread_local! {
     static TL_BUF: cell::RefCell<Vec<u8>> = cell::RefCell::new(Vec::with_capacity(128));
 }
 
+// Wrapper for `Write` types that counts total bytes written.
+struct CountingWriter<'a> {
+    wrapped: &'a mut io::Write,
+    count: usize,
+}
+
+impl<'a> CountingWriter<'a> {
+    fn new(wrapped: &'a mut io::Write) -> CountingWriter {
+        CountingWriter {
+            wrapped: wrapped,
+            count: 0,
+        }
+    }
+
+    fn count(&self) -> usize {
+        self.count
+    }
+}
+
+impl<'a> Write for CountingWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.wrapped.write(buf).map(|n| {
+            self.count += n;
+            n
+        })
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.wrapped.flush()
+    }
+
+    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+        self.wrapped.write_all(buf).map(|_| {
+            self.count += buf.len();
+            ()
+        })
+    }
+}
+
+type WriterFn = Fn(&mut io::Write) -> io::Result<()>;
+
+// Wrapper for `Write` types that executes a closure before writing anything,
+// but only if the write isn't empty. A `finish` call executes a closure after
+// writing, but again, only if something has been written.
+struct SurroundingWriter<'a> {
+    wrapped: &'a mut io::Write,
+    before: Option<&'a WriterFn>,
+    after: Option<&'a WriterFn>,
+}
+
+impl<'a> SurroundingWriter<'a> {
+    fn new(wrapped: &'a mut io::Write,
+           before: &'a WriterFn,
+           after: &'a WriterFn)
+           -> SurroundingWriter<'a> {
+        SurroundingWriter {
+            wrapped: wrapped,
+            before: Some(before),
+            after: Some(after),
+        }
+    }
+
+    fn do_before(&mut self, buf: &[u8]) -> io::Result<()> {
+        if buf.len() > 0 {
+            if let Some(before) = self.before.take() {
+                try!(before(self.wrapped));
+            }
+        }
+        Ok(())
+    }
+
+    fn finish(&mut self) -> io::Result<()> {
+        if let Some(after) = self.after.take() {
+            if self.before.is_none() {
+                try!(after(self.wrapped));
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<'a> Write for SurroundingWriter<'a> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        try!(self.do_before(buf));
```

Comment:
> This write would execute do_before multiple times if write or write_all are called multiple times, no? 

Yup, that's why I did `take()` on [line 102](https://github.com/slog-rs/term/pull/5/files#diff-8c13b5c367b1a0d3dc04da468ea844bdR102) :smile:

> It seems to me you could implement Drop for SurroundingWriter to do the after there so manually doing finish is not necessary

I thought about that too, and decided against it because then if there's an error writing `after`, you'll never get the error. I can add a `Drop` `impl` though too, and the `Drop` will just ignore the error.